ClausalBitvectorProof

alex-ozdemir opened the pull request:
## Summary

Created a class, `ClausalBitvectorProof`, which represents a bitvector
proof of UNSAT using an underlying clausal technique (DRAT, LRAT, etc)

It fits into the `BitvectorProof` class hierarchy like this:

```
              BitvectorProof
              /            \
             /              \
ClausalBitvectorProof  ResolutionBitvectorProof
```

## Why this is so nasty

This is probably going to be the nastiest diff in the DRAT/LRAT/ER sequence.

It is a painful one because all of the following BV subsystems
referenced ResolutionBitvectorProof (subsequently RBVP) or
BitvectorProof (subsequently BVP):
   * CnfStream
   * SatSolver (specifically the BvSatSolver)
   * CnfProof
   * TheoryProof
   * TheoryBV
   * Both bitblasters

And in particular, ResolutionBitvectorProof, the CnfStream, and the
SatSolvers were tightly coupled.

This means that references to and interactions with (R)BVP were
pervasive, and all of those references needed to be examined to figure out
which member of the new hierarchy they should refer to.

## Outline of changes

The change summary:
  * Create a subclass of BVP, called ClausalBitvectorProof, which has
    most methods stubbed out.
  * Make a some modifications to BVP and ResolutionBitvectorProof as the
    natural division of labor between the different classes becomes
    clear.
  * Go through all the components in the first list and try to figure
    out which kind of BVP they should **actually** be interacting with,
    and how. Make tweaks accordingly.
  * Add a hook from CryptoMinisat which pipes the produced DRAT proof
    into the new ClausalBitvectorProof.
  * Add a debug statement to ClausalBitvectorProof which parses and
    prints that DRAT proof, for testing purposes.

## Testing

Test:
  * `make check` to verify that we didn't break any old stuff, including
    lazy BB, and eager BB when using bvminisat.
  * `cvc4 --dump-proofs --bv-sat-solver=cryptominisat --bitblast=eager
  -d bv::clausal test/regress/regress0/bv/ackermann2.smt2`, and see that
     1. It crashed with "Unimplemented"
     2. Right before that it prints out the (textual) DRAT proof.

## Review
I assigned review to Andres, since he reviewed the BVP/RBVP class split, and this is a further articulation of that class hierarchy. However, if you think it would be better to have one the the BV developers review, feel free to re-assign.

4tXJ7f commented on src/proof/clausal_bitvector_proof.cpp:
```diff
@@ -0,0 +1,115 @@
+  int trueClauseId = ClauseId(ProofManager::currentPM()->nextId());
+  // with the CNF proof
+  d_cnfProof->registerTrueUnitClause(trueClauseId);
+  // and with (this) bitvector proof
```
```suggestion
  // and with (this) bit-vector proof
```

4tXJ7f commented on src/proof/clausal_bitvector_proof.cpp:
```diff
@@ -0,0 +1,115 @@
+void ClausalBitVectorProof::registerUsedClause(ClauseId id,
+                                               prop::SatClause& clause)
+{
+  d_usedClauses.push_back(std::make_pair(
```
Suggestion: could use `emplace_back()` instead.

alex-ozdemir replied:
Good idea!

alex-ozdemir commented on src/proof/bitvector_proof.h:
```diff
@@ -131,6 +137,28 @@ class BitVectorProof : public TheoryProof
+   * @param os the stream to print to
+   * @param paren any parentheses to add to the end of the global proof
+   */
+  virtual void printEmptyClauseProof(std::ostream& os, std::ostream& paren) = 0;
```
Sounds good! Will do.

alex-ozdemir replied:
Whoops! Yeah, it should be virtual. I'll fix that.

alex-ozdemir commented on src/proof/bitvector_proof.h:
```diff
@@ -153,13 +181,40 @@ class BitVectorProof : public TheoryProof
+   * high-level enough that it doesn't know the difference between clausal and
+   * resolution proofs.
+   *
+   * TODO(aozdemir) figure out what is going on and clean this up
```
Sounds good! Will do.

alex-ozdemir replied:
[here](https://github.com/CVC4/CVC4/issues/2789)

alex-ozdemir commented on src/proof/clausal_bitvector_proof.h:
```diff
@@ -0,0 +1,99 @@
+ **
+ ** \brief Bitvector proof using the DRAT & LRAT proof formats
+ **
+ ** An internal string stream is hooked up to cryptominisat, which spits out a
```
Whoops, this whole paragraph is somewhat out of date. I'll update it.

alex-ozdemir commented on src/proof/cnf_proof.cpp:
```diff
@@ -105,6 +105,30 @@ void CnfProof::registerConvertedClause(ClauseId clause, bool explanation) {
 
+void CnfProof::registerTrueUnitClause(ClauseId clauseId)
+{
+  Node true_node = NodeManager::currentNM()->mkConst<bool>(true);
```
Whoops! I'll fix it. I tend to accidentally use snake_case a lot.

alex-ozdemir commented on src/proof/cnf_proof.cpp:
```diff
@@ -105,6 +105,30 @@ void CnfProof::registerConvertedClause(ClauseId clause, bool explanation) {
+
+void CnfProof::registerFalseUnitClause(ClauseId clauseId)
+{
+  Node false_node = NodeManager::currentNM()->mkConst<bool>(false).notNode();
```
:+1:

alex-ozdemir commented on src/proof/resolution_bitvector_proof.h:
```diff
@@ -106,23 +85,24 @@ class ResolutionBitVectorProof : public BitVectorProof
 };
 
-class LFSCBitVectorProof : public ResolutionBitVectorProof
+class LFSCResolutionBitVectorProof : public ResolutionBitVectorProof
```
Good catch! I'll fix it!

alex-ozdemir commented on src/proof/theory_proof.cpp:
```diff
@@ -105,10 +117,11 @@ void TheoryProofEngine::finishRegisterTheory(theory::Theory* th) {
   if (th) {
     theory::TheoryId id = th->getId();
     if (id == theory::THEORY_BV) {
+      theory::bv::TheoryBV* bv_th = static_cast<theory::bv::TheoryBV*>(th);
```
Whoops! Thanks! I'll fix it.

alex-ozdemir commented on src/prop/cnf_stream.cpp:
```diff
@@ -84,14 +84,13 @@ void CnfStream::assertClause(TNode node, SatClause& c) {
+      d_cnfProof->registerConvertedClause(clause_id);
+    }
+    d_cnfProof->popCurrentDefinition();
+  });
```
Sounds good!

alex-ozdemir replied:
That is much better!

alex-ozdemir commented on src/prop/cryptominisat.cpp:
```diff
@@ -117,9 +118,17 @@ ClauseId CryptoMinisatSolver::addClause(SatClause& clause, bool removable){
-  d_okay &= res;
-  return ClauseIdError;
+  bool nowOkay = d_solver->add_clause(internal_clause);
+  int freshId = ClauseId(ProofManager::currentPM()->nextId());
```
Whoops! Yeah.

alex-ozdemir commented on src/theory/bv/bitblast/eager_bitblaster.cpp:
```diff
@@ -268,11 +268,12 @@ bool EagerBitblaster::collectModelInfo(TheoryModel* m, bool fullModel)
+  THEORY_PROOF(d_bvp = bvp; bvp->attachToSatSolver(*d_satSolver);
+               prop::SatVariable t = d_satSolver->trueVar();
+               prop::SatVariable f = d_satSolver->falseVar();
+               bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get(), t, f));
```
Good spot! Yeah, I should have seen that.

alex-ozdemir commented on src/theory/bv/bitblast/lazy_bitblaster.cpp:
```diff
@@ -566,11 +566,12 @@ bool TLazyBitblaster::collectModelInfo(TheoryModel* m, bool fullModel)
+  THEORY_PROOF(d_bvp = bvp; bvp->attachToSatSolver(*d_satSolver);
+               prop::SatVariable t = d_satSolver->trueVar();
+               prop::SatVariable f = d_satSolver->falseVar();
+               bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get(), t, f));
```
Agreed!

alex-ozdemir commented:
Thanks for the rapid review, Andres! I believe that I've addressed all of your comments.
If that's not the case, let me know. Also let me know if you think further work should be done to address any of them.

4tXJ7f commented:
Some regressions seem to be failing. Could you take a look?

alex-ozdemir commented:
Yeah, sorry, I'll do so, and report back when they're fixed.

alex-ozdemir commented:
Whoops -- I messed up the ordering of members within `TBitblaster`.

It turns out that the lazy and eager bitblasters ordered their (common) members differently. I hope that I have found an ordering that works for both!

alex-ozdemir commented:
Okay, it seems that this new ordering (the one from the lazy bitblaster) works!

alex-ozdemir commented:
ASAN reports some memory leaks. I don't _think_ they're related to this though.

4tXJ7f commented:
Sorry, I should have warned you. The leaks are a known issue. I actually just created a GitHub issue to track it (#2790). Removing the "do not merge" label.
